lib/tasks/grda_warehouse_clients.rake

```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+# rubocop:disable Naming/MethodParameterName
+
+require 'csv'
+
+namespace :grda_warehouse do
+  namespace :clients do
+    # rails grda_warehouse:clients:validate > warehouse_client_issues.csv
+    desc 'Report on validation issues in WarehouseClient associations, emits issues a csv'
+    task validate: [:environment] do
+      CSV do |csv|
+        csv << ['warehouse_client_id', 'error_message']
+        WarehouseClientValidator.validate do |wc, errors|
+          errors.each do |error|
+            csv << [wc.id, error]
+          end
+        end
+      end
+    end
+  end
+end
+
+class WarehouseClientValidator
+  def self.validate
+    destination_data_source_ids = GrdaWarehouse::DataSource.destination_data_source_ids.to_set
+    warehouse_clients = GrdaWarehouse::WarehouseClient.preload(:source, :destination)
+    destination_client_ids = GrdaWarehouse::WarehouseClient.distinct.pluck(:destination_id).to_set
+    warehouse_clients.find_each(batch_size: 1000) do |wc|
+      errors = validate_warehouse_client(
+        wc,
+        destination_data_source_ids,
+        destination_client_ids,
+      )
+      yield(wc, errors) if errors.any?
+    end
+  end
+
+  def self.validate_warehouse_client(wc, destination_data_source_ids, destination_client_ids)
+    errors = []
+
+    # Basic existence and equality checks
+    errors << "Source and destination are the same client (ID: #{wc.source_id})" if wc.source_id == wc.destination_id
+    errors << "Source client (ID: #{wc.source_id}) does not exist" unless wc.source
+    errors << "Destination client (ID: #{wc.destination_id}) does not exist" unless wc.destination
```

Comment:
Looks like we have missing source/destinations in QA. Perhaps these are expected?

Reply:
Not totally unexpected.  I think our cleanup of WarehouseClient records is minimal.  It might be a useful exercise to include a check against "deleted" clients and log the date they were marked deleted if available.

Reply:
I think client records have to exist due to the FK, so they must be just marked as deleted. Is there a reason to keep the WarehouseClient record around for a deleted client record? Or maybe we can delete them after some time, say `deleted_at < 1.month.ago` or something?

I could add another task, say `grda_warehouse:clients:prune` that deletes warehouse clients that seem like they could be safely removed. (if the source or destination client has been marked as deleted OR if the source_id == destination_id).